Making Sasl Bind async for advanced Sasl clients

Evengard opened the pull request:
Ideally, I believe all Bind methods should be async (or at least have an async variant), but for implementing some custom based Sasl clients which acts asynchronously (aka receive data to feed via EvaluateChallenge asynchronously - ie via a network, for example) the async is really a must have.
This might break a bit backward compatibility (because of interface changes), but I tried to make the port to be as harmless as possible (at least I hope). I kept all the old synchronized methods as is without changing their behaviour (I believe).

dsbenghe commented:
The async version of this library is implemented already in the "develop" branch. There I dropped completely the sync versions.

I haven't checked how your changes compare with the changes from the develop branch - maybe you can take a look to see if your changes can improve it?

Evengard commented:
I've fast checked your code, and I found out that your branch is missing a key element - it doesn't await for ISaslClient.EvaluateChallenge. This is basically the key element which I really needed and why I made this pull request.

I can try to port the changes to the develop branch.

dsbenghe commented:
Very true; your changes are only for bind sasl async which is not implemented in the develop branch. Don't seem to have conflicts with what I did so should be straight-forward to move them.

Evengard commented:
I've rebased the pull request against develop.
Now, I don't know your stance about the ISaslClient interface backward compatibility, but watching the other code in the develop branch I dropped the synchronous versions alltogether when rebasing the pull request. I can easily add it back though if it is needed.

dsbenghe commented:
My take on this: there is no problem to provide both sync and async versions as long as the sync version is not a wrapper of the async version i.e. if the sync version can not be provided without wrapping the async it is better to not provide it at all.

Evengard commented:
It seems that the pull request was reverted? What is the cause?
